WireDomain/Sources/WireDomain/Event Processing/ConversationEventProcessor/ConversationMLSMessageAddEventProcessor.swift

```diff
@@ -32,9 +34,133 @@ protocol ConversationMLSMessageAddEventProcessorProtocol {
+    let protobufMessageProcessor: any ConversationProtobufMessageProcessorProtocol
+
+    func processEvent(_ event: ConversationMLSMessageAddEvent) async throws {
+        let conversationID = event.conversationID
+        let senderID = event.senderID
+        let date = event.timestamp
+        let decryptedMessages = event.decryptedMessages
+        
+        guard !decryptedMessages.isEmpty else {
+            return WireLogger.proteus.warn(
+                "failed to add MLS message: there are no decrypted messages to process"
+            )
+        }
+
+        for decryptedMessage in decryptedMessages {
+            await processDecryptedMessage(
+                decryptedMessage,
+                conversationID: conversationID,
+                senderID: senderID,
+                date: date
+            )
+        }
+    }
+
+    private func processDecryptedMessage(
+        _ decryptedMessage: ConversationMLSMessageAddEvent.DecryptedMessage,
+        conversationID: ConversationID,
+        senderID: UserID,
+        date: Date?
+    ) async {
+        guard let conversation = await conversationLocalStore.fetchConversation(
+            id: conversationID.uuid,
+            domain: conversationID.domain
+        ) else {
+            return WireLogger.proteus.error(
+                "failed to add MLS message: conversation not found in db"
+            )
+        }
+
+        let logAttributes: LogAttributes = [
+            .messageType: "conversation.mls-message-add",
+            .conversationId: conversationID.uuid.safeForLoggingDescription
+        ]
+
+        // Ensure is self conversation, sender is self user and conversation is not read-only
+        guard await messageLocalStore.canAddMessage(
+            conversation: conversation,
+            senderID: senderID.uuid,
+            logAttributes: logAttributes
+        ) else {
+            return
+        }
+
+        // Read protobuf message
+        let protobufMessage = await readProtobufMessage(
+            from: decryptedMessage.message
+        )
+
+        guard let (genericMessage, content) = protobufMessage else {
+            WireLogger.eventProcessing.warn(
+                "Can't read protobuf, abort processing",
+                attributes: logAttributes
+            )
+
+            return await addInvalidSystemMessage(
+                senderID: senderID,
+                conversationID: conversationID,
+                date: date ?? .now
+            )
+        }
+
+        await conversationLocalStore.updateSecurityLevelAfterReceivingMessage(
+            conversation: conversation,
+            genericMessage: genericMessage,
+            date: date ?? .now
+        )
+
+        await conversationLocalStore.addParticipant(
+            participantID: senderID.uuid,
+            participantDomain: senderID.domain,
+            in: conversation,
+            date: date?.addingTimeInterval(-0.01) ?? .now
+        )
```

Comment:
I wonder why this is needed... can we receive messages from someone who we don't think is a participant yet?

Reply:
it seems that if the sender is not part of the conversation, it means we're out of sync, in that case we need to add the participant to the list, I renamed the method and added some doc comment so it is more clear: https://github.com/wireapp/wire-ios/pull/2164/commits/90da7addc7e950bb179e65ed21fe6583b88d9f4c